### PR TITLE
Enforce correct use of Serialize trait

### PR DESCRIPTION
**Pull request by dtolnay**
This ensures Serialize::serialize calls Serializer::serialize_* exactly once.

Half of #651.

### CODE REVIEW

**Comment by dtolnay** [comment]
JSON update: https://github.com/serde-rs/json/commit/ede8672b052070acbe968f4c71e355c89ce1b7db

**Comment by dtolnay** [comment]
I realized after implementing this that the `Serializer::Ok` type gives a really natural way to implement things like `serde_json::to_value`: https://github.com/serde-rs/json/commit/dc53c603b5fea6a49c68aca25183dd676a788e68

**Comment by dtolnay** [comment]
Merging so that I have a commit to tag as 0.9.0-rc1, but I will leave #651 open until @oli-obk gets a chance to look over this.

File: serde/src/ser/mod.rs [hunk 1]
```diff
@@ -77,345 +77,357 @@ pub trait Serialize {
 /// state object, as it is expected that the user will not modify it. Due to the generic nature
 /// of the `Serialize` impls, modifying the object is impossible on stable Rust.
 pub trait Serializer {
-    /// The error type that can be returned if some error occurs during serialization.
+    /// Trickery to enforce correct use of the `Serialize` trait. Every
+    /// `Serializer` should set `Ok = ()`.
+    type Ok;
```
Comment (oli-obk) [inline serde/src/ser/mod.rs]: I dislike this. But I can't think of another ergonomic-to-use way

Comment (dtolnay) [inline serde/src/ser/mod.rs]: Oh I should fix the comment. It actually turns out to be super useful. Check out https://github.com/serde-rs/json/commit/dc53c603b5fea6a49c68aca25183dd676a788e68.

Comment (oli-obk) [inline serde/src/ser/mod.rs]: Omfg, that is incredible!

**Pull request merged**